WalletWasabi/EventSourcing/InMemoryEventRepository.cs

```diff
@@ -0,0 +1,233 @@
+
+					// list of lists of events for atomic insertion of multiple events
+					// in one "database transaction"
+					ConcurrentQueue<IReadOnlyList<WrappedEvent>> EventsBatches
+				)
+			>
+		> _aggregatesEventsBatches = new();
+
+		private readonly ConcurrentDictionary<
+			string /*aggregateType*/,
+			(
+				// index of the last aggregateId in this aggregateType
+				long TailIndex,
+				ConcurrentDictionary<
+					// index of aggregateId in this aggregateType
+					long,
+					string /*aggregateId*/> Ids
+			)
+		> _aggregatesIds = new();
+
+		public Task AppendEventsAsync(
+			string aggregateType,
+			string aggregateId,
+			IEnumerable<WrappedEvent> wrappedEvents)
+		{
+			Guard.NotNullOrEmpty(nameof(aggregateType), aggregateType);
+			Guard.NotNullOrEmpty(nameof(aggregateId), aggregateId);
+			Guard.NotNull(nameof(wrappedEvents), wrappedEvents);
+			var wrappedEventsList = wrappedEvents.ToList().AsReadOnly();
+			if (wrappedEventsList.Count <= 0)
+			{
+				return Task.CompletedTask;
+			}
+			var firstSequenceId = wrappedEventsList[0].SequenceId;
+			var lastSequenceId = wrappedEventsList[^1].SequenceId;
+			if (firstSequenceId <= 0)
+			{
+				throw new ArgumentException("first event sequenceId is not natural number", nameof(wrappedEvents));
+			}
+			if (lastSequenceId <= 0)
+			{
+				throw new ArgumentException("last event sequenceId is not natural number", nameof(wrappedEvents));
+			}
+			if (lastSequenceId - firstSequenceId + 1 != wrappedEventsList.Count)
+			{
+				throw new ArgumentException("event sequence ids are out of whack", nameof(wrappedEvents));
+			}
+
+			var aggregateEventsBatches = _aggregatesEventsBatches.GetOrAdd(aggregateType, _ => new());
+			var (tailSequenceId, locked, eventsBatches) = aggregateEventsBatches.GetOrAdd(aggregateId, _ => (0, false, new()));
+
+			if (tailSequenceId + 1 < firstSequenceId)
+			{
+				throw new ArgumentException($"invalid firstSequenceId (gap in sequence ids) expected: '{tailSequenceId + 1}' given: '{firstSequenceId}'", nameof(wrappedEvents));
+			}
+
+			// no action
+			Validated();
+			// atomically detect conflict and replace lastSequenceId and lock to ensure strong order in eventsBatches
+			if (!aggregateEventsBatches.TryUpdate(
+				key: aggregateId,
+				newValue: (lastSequenceId, true, eventsBatches),
+				comparisonValue: (firstSequenceId - 1, false, eventsBatches)))
+			{
+				Conflicted(); // no action
+				throw new OptimisticConcurrencyException($"Conflict while commiting events. Retry command. aggregate: '{aggregateType}' id: '{aggregateId}'");
+			}
+			try
+			{
+				Locked(); // no action
+				eventsBatches.Enqueue(wrappedEventsList);
+				Appended(); // no action
+			}
+			finally
+			{
+				// unlock
+				if (!aggregateEventsBatches.TryUpdate(
+					key: aggregateId,
+					newValue: (lastSequenceId, false, eventsBatches),
+					comparisonValue: (lastSequenceId, true, eventsBatches)))
+				{
+#warning
+					// TODO: convert into Debug.Assert ???
```

Comment:
Isn't this more `InvalidOperationException` territory?

Reply:
nah, InvalidOperationException is basically input validation against internal state. This is just detection of broken algorithm.